rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,442 @@
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_connect_willmessage(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNECT(ref cv) = msg.op {
+            if let Some(p) = &cv.will_message {
+                if p.len() > 0 {
+                    unsafe {
+                        *buffer = p.as_ptr();
+                        *buffer_len = p.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
+    tx: &MQTTTransaction,
+    session_present: *mut bool,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNACK(ref ca) = msg.op {
+            unsafe {
+                *session_present = ca.session_present;
+            }
+            return 1;
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_topic(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.topic;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_message(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.message;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_topic(tx: &MQTTTransaction,
+                                       i: u16,
+                                       buf: *mut *const u8,
+                                       len: *mut u32)
+                                       -> u8
+{
+    // This function works on both SUBSCRIBE and UNSUBSCRIBE to reduce code
+    // duplication because both have multiple topics that need to be accessed
+    // via index from Rust. It does not imply that we expect a TX to have both
+    // SUBSCRIBE and UNSUBSCRIBE messages.
```

Comment:
Ah, I think I get your point just now: this function will always _only_ work on the first occurrence of either `SUBSCRIBE` or `UNSUBSCRIBE` and ignore the other. True, that is a problem. So someone can define a content match on the subscribe topic, but since they are both handled by the same code, it _might_ match on the unsubscribe topic if both are in there and it comes first.
I will address this by explicitly separating those in keywords _and_ Rust access functions.

Reply:
Ok, or you can keep one function and iterate over all topics of both `SUBSCRIBE` and `UNSUBSCRIBE`
You add a new variable `offset` initialize it to 0 and add an `else` to `if (i as usize) < subv.topics.len() {` where you increase `offset` by `subv.topics.len()`

Reply:
I'd rather split them, as it seems to be the cleaner solution here and reduces the potential for misunderstandings.

Reply:
ok for me :-)

Reply:
Done!